### DISCUSSION TITLE
Group permissions and invitation links

### DISCUSSION THREAD

**Discussion by aonghas** [pocketbase/pocketbase#4659] [Q&A] [answered]
Hi there,

My use case for invitations is slightly different from an old thread I commented on earlier, but I was wondering what the best solution would be for the following:

Users can create groups and each group can have a number of activities. Owners/admins of groups can add others to their group via an email field. Without knowing whether the person they're trying to add to the group is already a user of the app, my question was whether or not there was some way to generate an email to that user with a tokenized link. Clicking that link would redirect the user to some website page where either:

A) if they're already a user of the app and logged in, they can simply click a button to accept the group invitation

B) if they're a user of the app and not logged in, they could log in and then accept the invitation

C) if they're not a user of the app, they could sign up and then accept the invitation

I'm trying to figure out the best way to store the data to make API rules simple and reflect the different states of that process (eg. Before the user had accepted the invitation, the user in the app would be shown as "invitation sent" or something. Any pointers would be greatly appreciated!

**Comment by gokulgseven** [comment]
in the link you are sending send the group id as query paramaeter,
make a custom api that takes in the user id and the group id and adds that user to that group. (can be the default api too)
rest everything will be handeled with the front end

make a page in front end 
on load the page will check if the user is logged in or not 
   if the user is logged in display a button to trigger the above created api
   if user is not logged in 
       show the email and password field and a login button -> when the user loggs in , redirect back to the same page with the same group id in the quey parameter (hence you will now be displaying the butotn to join the group )
       on the same page display link to signup page -> on click redirect to that page again with the group id in the querry field 
       on the signup page display whatever fields you want for signup -> on click of the button signup the luser -> also login the user and redirect the user to the main page you created agin with the group if in the querry parameter 
       


basically this functionality is all done in front end and not something that pocketbase will help you with

**Reply by aonghas** [reply]
Cool thanks for that suggestion! I think it almost answers all my questions. A couple of questions/assumptions:

- When the group organiser first sends the invitation in the front end, how will Pocketbase handle the state before that user joins? I guess I need to hold that email somewhere in a table and then when the user clicks the accept invitation it can also resolve that pending record maybe through custom API...
- I guess your solution doesn't stop that same email invitation being forwarded on to anyone else? ie. it's an open invitation and the link would work for anyone, not just the specified user? Any ideas how to restrict this?

**Reply by gokulgseven** [reply]
yes it doesnt stop the same invite being used by someone else, 
if you want that store that in a table and map it to the particular user ,

**Reply by aonghas** [reply]
Just in case anyone else comes across this in future, my solution is as follows:

When an organiser wants to invite someone to join their group, a record is created in a separate collection called "invitations" which contains the invitee's email address, the group ID and a status "pending". This invitation record ID is then sent as a query parameterised link to the invitee.

The rest of the process follows similarly to what was suggested above by @gokulgseven except when the link is clicked in the email, it triggers a hook which simultaneously updates the invitation record status to "accepted" (or rejected) and adds that user ID to the group's members lookup field. This keeps the group API rules clean (since it's only updated server-side) and also means the email invitation link is one-time use. Also easier to create invitation expiration on the invitation record. 

Hope that's clear! Thanks all :)

**Comment by aonghas** [comment]
Update:

Am I doing this correctly? When I try to create a record in the invitations collection, I seem to get a 400 when trying to create a new invitation in the Admin UI so I'm thinking there's something wrong with the hook?

```
onRecordAfterCreateRequest((e) => {
  console.log(JSON.stringify(e.record));

  // get email template record from linked ID
  const templateRecord = $app
    .dao()
    .findRecordById("emailTemplates", e.record.template);

  // get group record from linked group
  const groupRecord = $app.dao().findRecordById("groups", e.record.groupId);

  const message = new MailerMessage({
    from: {
      address: $app.settings().meta.senderAddress,
      name: $app.settings().meta.senderName,
    },
    to: [{ address: e.record.email() }],
    subject: "You've been invited to: " + groupRecord.name,
    html: templateRecord.template
      .replace("{{groupName}}", groupRecord.name)
      .replace("{{groupId}}", groupRecord.id),
  });

  $app.newMailClient().send(message);
}, "invitations");
```

**Reply by ganigeorgiev** [reply, marked as answer]
`templateRecord.template` would be invalid (_`groupRecord.name` also_). Use the designated get/set methods to access the dynamic record fields - https://pocketbase.io/docs/js-records/#getset-record-fields.

If that's not the case check the logs for a more detailed error message in the _Admin UI > Logs_ and/or start the executable with `--dev`.
